pennylane/templates/state_preparations/state_prep_mps.py

```diff
@@ -498,3 +508,69 @@ def compute_decomposition(
     @MPSPrep._primitive.def_impl  # pylint: disable=protected-access
     def _(*args, **kwargs):
         return type.__call__(MPSPrep, args, **kwargs)
+
+
+def _mps_prep_decomposition_resources(bond_dimensions, num_sites, num_work_wires):  # pylint: disable=unused-argument
+    return {
+        resource_rep(qml.QubitUnitary, num_wires=1 + num_work_wires): num_sites
+    }
+
+
+def _work_wires_bond_dimension_condition(bond_dimensions, num_sites, num_work_wires):  # pylint: disable=unused-argument
+    max_bond_dimension = max(bond_dimensions[:-1])
+
+    return not ((max_bond_dimension > 2 ** num_work_wires) or (num_work_wires is not None and num_work_wires > 0))
+
+
+@qml.register_condition(_work_wires_bond_dimension_condition)
+@register_resources(_mps_prep_decomposition_resources)
+def _mps_prep_decomposition(*mps, **kwargs):
+    wires = kwargs["wires"]
+    work_wires = kwargs["work_wires"]
+    right_canonicalize = kwargs["right_canonicalize"]
+    mps = list(mps)
+
+    n_wires = len(work_wires) + 1
+
+    mps = mps.copy()
+
+    # Transform the MPS to ensure that the generated matrix is unitary
+    if right_canonicalize:
+        mps = right_canonicalize_mps(mps)
+
+    #  NOTE: tensor legs assignment convention is (vL, p, vR)
+    mps[0] = mps[0].reshape((1, *mps[0].shape))
+    mps[-1] = mps[-1].reshape((*mps[-1].shape, 1))
+
+    interface, dtype = qml.math.get_interface(mps[0]), mps[0].dtype
+
+    for i, Ai in enumerate(mps):
+
+        # Encode the tensor Ai in a unitary matrix following Eq.23 in https://arxiv.org/pdf/2310.18410
+        vectors = []
+        for column in Ai:
+            vector = qml.math.zeros(2**n_wires, like=interface, dtype=dtype)
+
+            if interface == "jax":
+                vector = vector.at[: len(column[0])].set(column[0])
+                vector = vector.at[2 ** (n_wires - 1) : 2 ** (n_wires - 1) + len(column[1])].set(
+                    column[1]
+                )
+
+            else:
+                vector[: len(column[0])] = column[0]
+                vector[2 ** (n_wires - 1) : 2 ** (n_wires - 1) + len(column[1])] = column[1]
+
+            vectors.append(vector)
+
+        vectors = qml.math.stack(vectors).T
+        # The unitary is completed using QR decomposition
+        d, k = vectors.shape
+        new_columns = qml.math.array(np.random.RandomState(42).random((d, d - k)))
```

Comment:
Do we want to be able to tune this randomness? @astralcai

Reply:
probably fine for now